File: src/com/gmail/marcosav2010/communicator/module/fth/command/FindCMD.java
package com.gmail.marcosav2010.communicator.module.fth.command;

import java.util.stream.Collectors;

import com.gmail.marcosav2010.command.Command;
import com.gmail.marcosav2010.communicator.module.fth.packet.PacketFindFile;
import com.gmail.marcosav2010.communicator.packet.wrapper.exception.PacketWriteException;
import com.gmail.marcosav2010.main.Main;
import com.gmail.marcosav2010.peer.ConnectedPeer;
import com.gmail.marcosav2010.peer.Peer;

class FindCMD extends Command {

    FindCMD() {
        super("find", "[peer] <filename>");
    }

    @Override
    public void execute(String[] arg, int args) {
        int pCount = Main.getInstance().getPeerManager().count();

        if (args < 1 && pCount == 1 || args < 2 && pCount > 1 || pCount == 0) {
            log.log("ERROR: Needed filename at least.");
            return;
        }

        boolean autoPeer = pCount > 1;
        Peer peer;

        if (!autoPeer) {
            String peerName = arg[0];

            if (Main.getInstance().getPeerManager().exists(peerName)) {
                peer = Main.getInstance().getPeerManager().get(peerName);
            } else {
                log.log("ERROR: Peer \"" + peerName + "\" doesn't exists.");
                return;
            }
        } else
            peer = Main.getInstance().getPeerManager().getFirstPeer();

        String filename = arg[autoPeer ? 0 : 1];

        log.log("Finding file \"" + filename + "\"...");

        var connectedPeers = peer.getConnectionManager().getIdentificator().getConnectedPeers();
        var p = new PacketFindFile(filename, 1,
                connectedPeers.stream().map(ConnectedPeer::getUUID).collect(Collectors.toSet()));

        connectedPeers.forEach(c -> {
            try {
                c.sendPacket(p);
            } catch (PacketWriteException e) {
                log.log(e, "There was a problem while sending find packet to " + c.getName() + ".");
            }
        });
    }
}

File: modules/file-transfer/src/main/java/com/gmail/marcosav2010/fth/command/FindCMD.kt
package com.gmail.marcosav2010.fth.command

import com.gmail.marcosav2010.command.Command
import com.gmail.marcosav2010.communicator.packet.wrapper.exception.PacketWriteException
import com.gmail.marcosav2010.fth.packet.PacketFindFile
import com.gmail.marcosav2010.main.Main
import com.gmail.marcosav2010.peer.Peer

internal class FindCMD : Command("find", "[peer] <filename>") {

    override fun execute(arg: Array<String>, length: Int) {
        val peerManager = Main.instance.peerManager!!

        val pCount = peerManager.count()
        if (length < 1 && pCount == 1 || length < 2 && pCount > 1 || pCount == 0) {
            log.log("ERROR: Needed filename at least.")
            return
        }

        val autoPeer = pCount > 1
        val peer: Peer = if (!autoPeer) {
            val peerName = arg[0]
            if (peerManager.exists(peerName)) {
                peerManager[peerName]!!
            } else {
                log.log("ERROR: Peer \"$peerName\" doesn't exists.")
                return
            }
        } else peerManager.firstPeer

        val filename = arg[if (autoPeer) 0 else 1]
        log.log("Finding file \"$filename\"...")

        val connectedPeers = peer.connectionManager.identificator.connectedPeers
        val p = PacketFindFile(filename, 1, connectedPeers.map { obj -> obj.uuid }.toMutableSet())

        connectedPeers.forEach { c ->
            try {
                c.sendPacket(p)
            } catch (e: PacketWriteException) {
                log.log(e, "There was a problem while sending find packet to " + c.name + ".")
            }
        }
    }
}